STABILITY.md

```diff
@@ -0,0 +1,123 @@
+  - A crate which depends on a crate in this workspace directly.
+* _downstream crate_
+  - A crate which directly or indirectly depends on a crate in this workspace.
+* _workspace crate_
+  - A crate which is a member of this workspace and hence is stored in this repository
+* _breaking change_
+  - A change in code that requires a `dependent crate` to adjust their code to fix compile errors.
+* _release_
+  - A new version of a crate is published to crates.io 
+
+## Tiers
+
+The project uses three stability tiers for all of its crates, and all crates use [semver] for their version numbers.
+Tiers differ primarily in the time between breaking changes, which always have to be announced with `PRs` as per
+our [collaboration guide].
+
+The following schematic helps to visualize what follows.
+
+```
+                Release Software v1.X              
+    Stability Tier 1 ═════════════════════════════╗
+    ║                                             ║
+    ║    gixp─────────────┐ gix──────────────┐    ║
+    ║    │  plumbing app  │ │  porcelain app │    ║
+    ║    └────────────────┘ └────────────────┘    ║
+    ║             │                  │            ║
+    ║             ▼                  ▼            ║
+    ║    gitoxide-core───────────────────────┐    ║
+    ║    │        application functionality  │    ║
+    ║    └───────────────────────────────────┘    ║
+    ║                      │                      ║
+    ║                      ▼                      ║
+    ║    git-repository──────────────────────┐    ║
+    ║    │                application crate  │    ║
+    ║    └───────────────────────────────────┘    ║
+    ║                      │                      ║
+    ╚═════════════════════════════════════════════╝
+                           │                       
+    Stability Tier 2 ─────────────────────────────┐
+    │                      ▼                      │
+    │    plumbing crates─────────────────────┐    │
+    │    │ ┌───────────┐       ┌───────────┐ │    │
+    │    │ │  git-odb  │       │  git-ref  │ │    │
+    │    │ └───────────┘       └───────────┘ │    │
+    │    │ ┌───────────┐       ┌───────────┐ │    │
+    │    │ │git-config │       │ git-pack  │ │    │
+    │    │ └───────────┘       └───────────┘ │    │
+    │    │ ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐ │    │
+    │    │            …many more…            │    │
+    │    │ └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘ │    │
+    │    └───────────────────────────────────┘    │
+    └─────────────────────────────────────────────┘
+ ```
+
+### Tier 3: pre-release crates
+
+Pre-release crates is marked with major version number zero, for example `0.1.4` and lives in stability tier 3 _(->ST3)_.
+
+It's acceptable to let each breaking change be followed by a minor version release.
+Dependent crates are advised to use `=<version>` version requirements to avoid automatic updates of patch levels which
+may be breaking if they also depend on other `plumbing crates`.
```

Comment:
Is this necessary? This is implied by semver, and fwiw is the convention in Rust's ecosystem for pre 1.0 crates anyways.

Reply:
I think it's beneficial to spell out how this is handled even if it represents the status quo.

Also I find helpful to explicitly state that dependents should pin their dependencies for stability. For the latter to truly work `gitoxide` crates should also pin their dependencies, so that A-0.4 -> B 0.3 won't have a chance to break downstream if A-0.4.1 -> B -> 0.4 is released with dependent D using A (which auto-updates) and B 0.3, which won't auto-update.

Reply:
> Also I find helpful to explicitly state that dependents should pin their dependencies for stability. For the latter to truly work gitoxide crates should also pin their dependencies, so that A-0.4 -> B 0.3 won't have a chance to break downstream if A-0.4.1 -> B -> 0.4 is released with dependent D using A (which auto-updates) and B 0.3, which won't auto-update.

Please ignore the latter as I think it's not necessary to protect downstream from that possibility with pre-release crates. This should only be done with 'alpha' releases of released software to not create issues like the ones encountered recently with `clap`.